profile_qadi_performance.py

```
@@ -0,0 +1,335 @@
+            potential_parallel_time = max(phase_durations)
+            print(f"  - Potential time savings with parallelization: {sequential_time - potential_parallel_time:.2f}s")
+    
+    # Agent setup time
+    setup_events = [e for e in report["detailed_events"] if "setup" in e["name"].lower()]
+    if setup_events:
+        total_setup_time = sum(e["duration"] for e in setup_events)
+        print(f"  - Agent setup time: {total_setup_time:.2f}s ({total_setup_time/report['total_duration']*100:.1f}% of total)")
+    
+    # Save detailed report
+    report_path = "/Users/yuyamukai/dev/mad_spark_alt/performance_report.json"
+    with open(report_path, "w") as f:
```

Ruff (0.11.9)

Async functions should not open files with blocking methods like `open`

(ASYNC230)